### PR TITLE
Updates to `feature/egressExtension` Part 1

### PR DESCRIPTION
**Pull request by kkeirstead**
###### Summary

Part 1 of my updates to the `feature/egressExtension` branch - I pulled out some things to keep this PR smaller, as well as areas that I'm still working through. I'll leave comments throughout the code to explain different things that I think are worth pointing out or that I've cut out (for now). Builds on top of https://github.com/dotnet/dotnet-monitor/pull/1953.

### CODE REVIEW

File: src/Extensions/AzureBlobStorage/EgressArtifactSettings.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        /// <summary>
+        /// The metadata of the blob to be created.
+        /// </summary>
+        public Dictionary<string, string> Metadata { get; set; }
+            = new Dictionary<string, string>(StringComparer.Ordinal);
+
+        /// <summary>
+        /// Custom metadata of the blob to be created.
+        /// </summary>
+        public Dictionary<string, string> CustomMetadata { get; }
```
Comment (kkeirstead) [inline src/Extensions/AzureBlobStorage/EgressArtifactSettings.cs:23]: Not currently being populated (I'll mention more about this in another comment)

File: src/Microsoft.Diagnostics.Monitoring.Options/ExtensionEgressProviderOptions.cs [hunk 2]
```diff
@@ -5,15 +5,14 @@
...
-using System.Text.Json.Serialization;
 
 namespace Microsoft.Diagnostics.Tools.Monitor.Egress
 {
     /// <summary>
     /// Egress provider options for external egress providers.
     /// </summary>
     internal sealed partial class ExtensionEgressProviderOptions : 
-        Dictionary<string, string>,
+        Dictionary<string, object>,
```
Comment (kkeirstead) [inline src/Microsoft.Diagnostics.Monitoring.Options/ExtensionEgressProviderOptions.cs:15]: This isn't implemented in this PR, but in the initial iteration of this design we were assuming that we could treat all of our extension's configuration as a `Dictionary<string, string>`. The addition of the Azure provider's `Metadata` (used for custom metadata) breaks this assumption, and will require changes to how we pass the configuration to the extension. To simplify this PR, I am not currently passing the `Metadata` section of configuration, but this will be fixed in a future PR.

File: src/Tools/dotnet-monitor/Egress/Configuration/EgressProviderConfigureNamedOptions.cs [hunk 3]
```diff
@@ -34,7 +34,23 @@ public void Configure(string name, TOptions options)
                 IConfigurationSection providerOptionsSection = providerTypeSection.GetSection(name);
                 if (providerOptionsSection.Exists())
                 {
-                    providerOptionsSection.Bind(options);
+                    if (options is ExtensionEgressProviderOptions extensionOptions)
```
Comment (kkeirstead) [inline src/Tools/dotnet-monitor/Egress/Configuration/EgressProviderConfigureNamedOptions.cs:37]: This is where we'll need to handle `Metadata` (and anything else that doesn't have a value for a child).

File: src/Tools/dotnet-monitor/Extensibility/FolderExtensionRepository.cs [hunk 4]
```diff
@@ -28,15 +29,42 @@ public FolderExtensionRepository(IFileProvider fileSystem, ILoggerFactory logger
 
         public override bool TryFindExtension(string extensionName, out IExtension extension)
         {
-            IDirectoryContents extensionDir = _fileSystem.GetDirectoryContents(extensionName);
+            string extensionPath = string.Empty;
+
+            bool isDotnetToolsLocation = _targetFolder == HostBuilderSettings.ExtensionDirectoryPath;
+
+            // Special case -> need to look in particular path for dotnet tool installation
```
Comment (kkeirstead) [inline src/Tools/dotnet-monitor/Extensibility/FolderExtensionRepository.cs:36]: This logic is for the `dotnet tool install` scenario, so that users don't need to manually place an extension downloaded from Nuget into the right directory. This works in my limited testing (building a nuget package locally, and then calling `dotnet tool install --add-source...` on it), but I think this area may need some more refinement going forward.

File: src/Tools/dotnet-monitor/LoggingEventIds.cs [hunk 5]
```diff
@@ -66,10 +65,7 @@ internal enum LoggingEventIds
         LoadingProfiler = 53,
         SetEnvironmentVariable = 54,
         GetEnvironmentVariable = 55,
-        MonitorApiKeyNotConfigured = 56,
-        QueueDoesNotExist = 57,
-        QueueOptionsPartiallySet = 58,
-        WritingMessageToQueueFailed = 59,
+        MonitorApiKeyNotConfigured = 56, // Note the gap - from removing things related to Azure egress
```
Comment (kkeirstead) [inline src/Tools/dotnet-monitor/LoggingEventIds.cs:68]: Will we want to update all of these numbers to be in-order again? I haven't for now, but we have several gaps left behind by the logging specific to the Azure extension.

File: src/Extensions/AzureBlobStorage.sln [hunk 6]
```diff
@@ -0,0 +1,25 @@
+﻿
```
Comment (jander-msft) [inline src/Extensions/AzureBlobStorage.sln:1]: Remove this and add the AzureBlobStorage project to the dotnet-monitor.sln file if it already doesn't exist.

File: src/Extensions/AzureBlobStorage/AzureBlob/AzureBlobEgressProvider.cs [hunk 7]
```diff
@@ -9,36 +9,28 @@
...
     /// </summary>
     /// <remarks>
     /// Blobs created through this provider will overwrite existing blobs if they have the same blob name.
     /// </remarks>
-    internal partial class AzureBlobEgressProvider :
-        EgressProvider<AzureBlobEgressProviderOptions>
+    internal partial class AzureBlobEgressProvider
     {
-        private int BlobStorageBufferSize = 4 * 1024 * 1024;
+        private readonly string AzureBlobStorage = "AzureBlobStorage";
```
Comment (jander-msft) [inline src/Extensions/AzureBlobStorage/AzureBlob/AzureBlobEgressProvider.cs:25]: This should be moved to some kind of shared constant because it is used in the same way both here and in the Program class.

File: src/Extensions/AzureBlobStorage/AzureBlob/AzureBlobEgressProvider.cs [hunk 8]
```diff
@@ -9,36 +9,28 @@
...
+        protected ILogger Logger { get; }
```
Comment (jander-msft) [inline src/Extensions/AzureBlobStorage/AzureBlob/AzureBlobEgressProvider.cs:26]: This should be a private field.

File: src/Extensions/AzureBlobStorage/Program.cs [hunk 9]
```diff
@@ -2,38 +2,43 @@
...
         private static Stream StdInStream = null;
         private static CancellationTokenSource CancelSource = new CancellationTokenSource();
+
+        protected static ILogger Logger { get; set; }
+
         static async Task<int> Main(string[] args)
         {
-            // Expected command line format is: AzureBlobStorage.exe Egress --Provider-Name MyProviderEndpointName
-            RootCommand rootCommand = new RootCommand("Egresses an artifact to Azure Storage.");
+            using var loggerFactory = LoggerFactory.Create(builder =>
```
Comment (jander-msft) [inline src/Extensions/AzureBlobStorage/Program.cs:21]: This should be done in the Egress method and not stored as static state on the Program class.

File: src/Extensions/CODE-OF-CONDUCT.md [hunk 10]
```diff
@@ -0,0 +1,6 @@
+# Code of Conduct
```
Comment (jander-msft) [inline src/Extensions/CODE-OF-CONDUCT.md:1]: Do not create a separate file here. This is still part of the dotnet-monitor repo so it should be using the files at the root.

File: src/Extensions/LICENSE.TXT [hunk 11]
```diff
@@ -0,0 +1,23 @@
+The MIT License (MIT)
```
Comment (jander-msft) [inline src/Extensions/LICENSE.TXT:1]: Do not create a separate license file here. This is still part of the dotnet-monitor repo so it should be using the files at the root.

**Comment by kkeirstead** [comment]
Closing this PR in favor of breaking things down even further for ease of review. I will incorporate the feedback already left on this PR into my changes.

File: src/Tools/dotnet-monitor/Egress/Configuration/EgressProviderConfigureNamedOptions.cs [hunk 12]
```diff
@@ -34,7 +34,23 @@ public void Configure(string name, TOptions options)
...
+                    {
+                        var children = providerOptionsSection.GetChildren();
```
Comment (jander-msft) [inline src/Tools/dotnet-monitor/Egress/Configuration/EgressProviderConfigureNamedOptions.cs:39]: Why is this needed? Are you working around this issue? https://github.com/dotnet/runtime/issues/77246

File: src/Tools/dotnet-monitor/Egress/Extension/ExtensionEgressProvider.cs [hunk 13]
```diff
@@ -43,8 +42,7 @@ public ExtensionEgressProvider(IEgressPropertiesProvider propertyProvider, Exten
             {
                 Settings = artifactSettings,
                 Configuration = options,
-                Properties = _propertyProvider.GetAllProperties(),
-                ProviderName = providerName,
```
Comment (jander-msft) [inline src/Tools/dotnet-monitor/Egress/Extension/ExtensionEgressProvider.cs:47]: Why is the provider name being removed?

File: src/Tools/dotnet-monitor/LoggingEventIds.cs (hunk 5 above)
Comment (jander-msft) [reply src/Tools/dotnet-monitor/LoggingEventIds.cs:68]: Do not remove or reorder anything here. Just add a comment after then that they've moved to the extension.

File: src/Tools/dotnet-monitor/HostBuilder/HostBuilderSettings.cs [hunk 14]
```diff
@@ -42,6 +44,16 @@ internal sealed class HostBuilderSettings
                     Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), "." + ProductFolderName) :
                     Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), ProductFolderName));
 
+        // Location where extensions are stored by default.
+        // Windows: "%USERPROFILE%\.dotnet\Tools"
+        // Other: "%XDG_CONFIG_HOME%/dotnet/tools" OR "%HOME%/.config/dotnet/tools" -> IS THIS RIGHT?
```
Comment (jander-msft) [inline src/Tools/dotnet-monitor/HostBuilder/HostBuilderSettings.cs:49]: It appears to be the same as on Windows:

```bash
jander@DESKTOP-N8NH8I6:~/.dotnet/tools$ ls
dotnet-dump  dotnet-monitor  dotnet-sos  dotnet-trace
```

File: src/Tools/dotnet-monitor/Extensibility/FolderExtensionRepository.cs [hunk 15]
```diff
@@ -28,15 +29,42 @@ public FolderExtensionRepository(IFileProvider fileSystem, ILoggerFactory logger
 
         public override bool TryFindExtension(string extensionName, out IExtension extension)
         {
-            IDirectoryContents extensionDir = _fileSystem.GetDirectoryContents(extensionName);
+            string extensionPath = string.Empty;
+
+            bool isDotnetToolsLocation = _targetFolder == HostBuilderSettings.ExtensionDirectoryPath;
```
Comment (jander-msft) [inline src/Tools/dotnet-monitor/Extensibility/FolderExtensionRepository.cs:34]: I think you should have a separate ExtensionRepository implementation for the dotnet tools folder since its algorithm is very much different than the FolderExtensionRepository implementation.
